### PR TITLE
make submap() handle both 'BaseCoordinateFrame' and 'SkyCoord'

### PR DESCRIPTION
**Pull request by OussCHE**
## PR Description

As mentioned in the discussion of the #6252 issue and following the advice of @ayshih ,I modified the function so that `submap()` can handle `BaseCoordinateFrame` objects. 

However, I wanted to add a test to make sure that the functionality results in the same `AIAMAP` object,but I didn't find a way to compare the resulting objects (I compared them in debugging and using the code bellow, they look the same for me).

I wrote a snippet code that should do a big part of the job in comparing between the results(and to be used in the test). I can write the test(or modify my code) if you give me some feedback.

    import sunpy.map
    from sunpy.data.sample import AIA_171_IMAGE
    from astropy.coordinates import SkyCoord
    import numpy as np
    aia = sunpy.map.Map(AIA_171_IMAGE)
    bottom_left = aia.bottom_left_coord.frame
    top_right = aia.top_right_coord.frame
    basecoordianteframe_value = aia.submap(bottom_left, top_right=top_right)

    bottom_left =SkyCoord(aia.bottom_left_coord.frame)
    top_right = SkyCoord(aia.top_right_coord.frame)
    skycoord_value = aia.submap(bottom_left, top_right=top_right)
    
    np.testing.assert_equal(basecoordianteframe_value.data,skycoord_value.data)

Cheers

### CODE REVIEW

**Comment by nabobalis** [comment]
Thank you for the PR, I think we will need a unit test to make sure this works and a bugix changelog entry:  https://github.com/sunpy/sunpy/blob/main/changelog/README.rst

**Comment by OussCHE** [comment]
@nabobalis sure, I will add a similar test to take into account that `submap()` runs with `BaseCoordinateFrame` with no issues. I will also check on the changelog

**Comment by OussCHE** [comment]
@nabobalis I have just added a test and a changelog for the bug.

File: changelog/6820.bugfix.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+make submap() handle both 'BaseCoordinateFrame' and 'SkyCoord' objects
```
Comment (nabobalis) [inline changelog/6820.bugfix.rst:1]:
```suggestion
:meth:`sunpy.map.GenericMap.submap` can now handle a '~astropy.coordinates.BaseCoordinateFrame' as input.
```

File: sunpy/map/tests/test_mapbase.py [hunk 2]
```diff
@@ -1603,3 +1603,11 @@ def test_only_cd():
     cd_map = sunpy.map.Map((data, header))
     np.testing.assert_allclose(u.Quantity(cd_map.scale).value, np.array([5, 13]))
     np.testing.assert_allclose(cd_map.rotation_matrix, np.array([[3/5, -4/5], [5/13, 12/13]]))
+
+
+def test_basecoordianteframe_submap(aia171_test_map):
```
Comment (nabobalis) [inline sunpy/map/tests/test_mapbase.py:1608]: Is there a better place to for this test to other than at the end of this file?

File: sunpy/map/tests/test_mapbase.py [hunk 3]
```diff
@@ -1603,3 +1603,11 @@ def test_only_cd():
...
+    #test that submap runs with BaseCoordinateFrame objects
```
Comment (nabobalis) [inline sunpy/map/tests/test_mapbase.py:1609]:
```suggestion
    # Tests the fix for https://github.com/sunpy/sunpy/issues/6252
```

**Review by nabobalis — COMMENTED** [review]
The style problems (pre-commit check) also needs addressing.

File: sunpy/map/tests/test_mapbase.py (hunk 2 above)
Comment (OussCHE) [reply sunpy/map/tests/test_mapbase.py:1608]: I tried to find similar tests and put them besides them. There are some tests that have the same parameter,but they don't look like similar tests

Comment (OussCHE) [reply sunpy/map/tests/test_mapbase.py]: I added the modifications @ayshih

**Comment by nabobalis** [comment]
The code style (pre-commit check) will need addressing.

**Comment by nabobalis** [comment]
> @nabobalis should I worry about Codecov?

No, the tests did not run. So there was nothing to report.

**Review by ayshih — APPROVED** [review]
Looks good to me.  Thanks, @OussCHE!

Fixes #6252 

Squash upon merging

**Comment by nabobalis** [comment]
Thank you @OussCHE for the PR and thank you @ayshih for the review.

**Pull request merged**